modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/MergeBehavior.java

```diff
@@ -170,8 +98,143 @@ public RenderedImage process(
+                        backgroundValues,
+                        inputThreshold,
+                        sourceAlpha,
+                        sourceROI,
+                        mosaicType,
+                        localHints);
+            }
+
+            // put in the same envelope
+            sources =
+                    harmonizeSources(
+                            sources,
+                            backgroundValues,
+                            inputThreshold,
+                            sourceAlpha,
+                            sourceROI,
+                            localHints);
+
+            return new ImageWorker(localHints).setImage(sources[0]).min(sources).getRenderedImage();
+        }
     };
 
+    /**
+     * Prepares multiple sources, potentially not sharing the same extent, for an operation that
+     * requires an exact overlap.
+     */
+    private static RenderedImage[] harmonizeSources(
+            RenderedImage[] sources,
+            double[] backgroundValues,
+            double[][] inputThreshold,
+            PlanarImage[] sourceAlpha,
+            ROI[] sourceROI,
+            RenderingHints hints) {
+        RenderedImage[] result = Arrays.copyOf(sources, sources.length);
+
+        // Check if the images don't share the same bounds
+        Rectangle union = new Rectangle(PlanarImage.wrapRenderedImage(sources[0]).getBounds());
+        boolean performMosaic = false;
+        for (int i = 1; i < sources.length; i++) {
+            Rectangle currentExtent = PlanarImage.wrapRenderedImage(sources[i]).getBounds();
+            if (!currentExtent.equals(union)) {
+                performMosaic = true;
+                union = union.union(currentExtent);
+            }
+        }
+        if (!performMosaic) return result;
+
+        // different extents found, use mosaic to extend each one of them to the union of the bounds
+        // shared image layout for all images
+        final ImageLayout layout =
+                (ImageLayout)
+                        (hints != null ? hints.get(JAI.KEY_IMAGE_LAYOUT) : new ImageLayout2());
+        layout.setWidth(union.width).setHeight(union.height).setMinX(union.x).setMinY(union.y);
+
+        final RenderingHints localHints;
+        if (hints != null) {
+            localHints = (RenderingHints) hints.clone();
+            localHints.add(new RenderingHints(JAI.KEY_IMAGE_LAYOUT, layout));
+        } else {
+            localHints = new RenderingHints(JAI.KEY_IMAGE_LAYOUT, layout);
+        }
+
+        // extend each one using mosaic
+        for (int i = 0; i < sources.length; i++) {
+
+            PlanarImage[] sourceAlphas =
+                    sourceAlpha == null ? null : new PlanarImage[] {sourceAlpha[i]};
+            ROI[] sourceROIs = sourceROI == null ? null : new ROI[] {sourceROI[i]};
+            Range nodata = new ImageWorker(sources[i]).getNoData();
+            Range[] nodatas = nodata == null ? null : new Range[] {nodata};
```

Comment:
Where the nodata is being used?

Reply:
Oh right, it's not, I should remove it. It's a bit crazy, but the integration of mosaic is using the input nodata only if the nodata is left null, otherwise the nodata being provided is used as background values and stuff goes belly up. I could not make sense of what's going on, but the simplest fix, is to remove this line.